accounts-db/src/io_uring/file_creator.rs

```diff
@@ -0,0 +1,550 @@
+struct CloseOp {
+    file_key: usize,
+}
+
+impl<'a> CloseOp {
+    fn new(file_key: usize) -> Self {
+        Self { file_key }
+    }
+
+    fn entry(&mut self) -> squeue::Entry {
+        opcode::Close::new(types::Fixed(self.file_key as u32)).build()
+    }
+
+    fn complete(
+        &mut self,
+        ring: &mut Completion<FileCreatorState<'a>, FileCreatorOp>,
+        res: io::Result<i32>,
+    ) -> io::Result<()>
+    where
+        Self: Sized,
+    {
+        let _ = res?;
+        ring.context_mut().mark_file_closed(self.file_key);
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct WriteOp {
+    file_key: usize,
+    offset: usize,
+    buf: IoFixedBuffer,
+    write_len: usize,
+}
+
+impl<'a> WriteOp {
+    fn entry(&mut self) -> squeue::Entry {
+        let WriteOp {
+            file_key,
+            offset,
+            buf,
+            write_len,
+        } = self;
+
+        opcode::WriteFixed::new(
+            types::Fixed(*file_key as u32),
+            buf.as_mut_ptr(),
+            *write_len as u32,
+            buf.io_buf_index()
+                .expect("should have a valid fixed buffer"),
+        )
+        .offset(*offset as u64)
+        .ioprio(IO_PRIO_BE_HIGHEST)
+        .build()
+        .flags(squeue::Flags::ASYNC)
+    }
+
+    fn complete(
+        &mut self,
+        ring: &mut Completion<FileCreatorState<'a>, FileCreatorOp>,
+        res: io::Result<i32>,
+    ) -> io::Result<()>
+    where
+        Self: Sized,
+    {
+        let written = res? as usize;
+
+        let WriteOp {
+            file_key,
+            offset: _,
+            ref mut buf,
+            write_len,
+        } = self;
+
+        assert_eq!(written, *write_len, "short write");
+
+        let buf = mem::replace(buf, IoFixedBuffer::empty());
+        if ring
+            .context_mut()
+            .mark_write_completed(*file_key, *write_len, buf)
+        {
+            ring.push(FileCreatorOp::Close(CloseOp::new(*file_key)));
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+enum FileCreatorOp {
+    Open(OpenOp),
+    Close(CloseOp),
+    Write(WriteOp),
+}
+
+impl RingOp<FileCreatorState<'_>> for FileCreatorOp {
+    fn entry(&mut self) -> squeue::Entry {
+        match self {
+            Self::Open(op) => op.entry(),
+            Self::Close(op) => op.entry(),
+            Self::Write(op) => op.entry(),
+        }
+    }
+
+    fn complete(
+        &mut self,
+        ring: &mut Completion<FileCreatorState, Self>,
+        res: io::Result<i32>,
+    ) -> io::Result<()>
+    where
+        Self: Sized,
+    {
+        match self {
+            Self::Open(op) => op.complete(ring, res),
+            Self::Close(op) => op.complete(ring, res),
+            Self::Write(op) => op.complete(ring, res),
+        }
+    }
+}
+
+type PendingWrite = (IoFixedBuffer, usize, usize);
+
+#[derive(Debug)]
+struct PendingFile {
+    path: PathBuf,
+    completed_open: bool,
+    backlog: SmallVec<[PendingWrite; 8]>,
```

Comment:
I remember writing this backlog thing, but I don't remember where 8 comes from?
Have you measured?